File: hack/docgen.py
import json
import re
from pathlib import Path

SECTION_HEADER = """

# %s
"""

FIELD_HEADER = """

## %s

%s"""

FIELD_TABLE_HEADER = """

### Fields
| Field Name | Field Type | Description   |
|:----------:|:----------:|---------------|"""

TABLE_ROW = """
|`%s`|%s|%s|"""

DEP_TABLE_ROW = """
|~`%s`~|~%s~|%s|"""

DROPDOWN_OPENER = """
<details>
<summary>%s (click to open)</summary>
<br>"""

LIST_ELEMENT = """

- %s"""

DROPDOWN_CLOSER = """
</details>"""


def clean_title(title):
    if "\n+" in title:
        return title[:title.find("\n+")]
    return title


def clean_desc(desc):
    desc = desc.replace("\n", "")
    dep = ""
    if 'DEPRECATED' in desc:
        dep = " " + desc[desc.find("DEPRECATED"):]

    if '+patch' in desc:
        desc = desc[:desc.find("+patch")]
    if '+proto' in desc:
        desc = desc[:desc.find("+proto")]
    if '+option' in desc:
        desc = desc[:desc.find("+option")]

    if dep != "" and 'DEPRECATED' not in desc:
        desc += dep
    return desc


def get_row(name, _type, desc):
    if 'DEPRECATED' in desc:
        dep = desc.find('DEPRECATED')
        return DEP_TABLE_ROW % (name, _type, "~" + desc[:dep] + "~ " + desc[dep:])
    else:
        return TABLE_ROW % (name, _type, desc)


def get_name_from_full_name(ref):
    return ref.split('.')[-1]


def get_desc(key, defs):
    if 'description' in defs[key]:
        return clean_desc(defs[key]['description'])
    elif 'title' in defs[key]:
        return clean_desc(clean_title(defs[key]['title']))
    return "_No description available_"


def link(text, link_to):
    return "[%s](%s)" % (text, link_to)


def get_desc_from_field(field):
    if 'description' in field:
        return field['description']
    elif 'title' in field:
        return field['title']
    return "_No desription available_"


def get_examples(examples, summary):
    out = DROPDOWN_OPENER % summary
    for example in examples:
        name = example.split("/")[-1]
        out += LIST_ELEMENT % link("`%s`" % name, "../" + example)
    out += DROPDOWN_CLOSER
    return out


def get_key_value_field_types(field):
    key_type, val_type = "string", "string"
    if 'type' in field['additionalProperties']:
        key_type = field['additionalProperties']['type']
    if 'format' in field['additionalProperties']:
        val_type = field['additionalProperties']['format']
    return key_type, val_type


def get_object_type(field, json_field_name, add_to_queue):
    obj_type_raw = field['type']
    if obj_type_raw == "array":
        if '$ref' in field['items']:
            ref = field['items']['$ref'][14:]
            add_to_queue(ref, json_field_name)

            if ref == "io.argoproj.workflow.v1alpha1.WorkflowStep":
                return "`Array<Array<`%s`>>`" % link("`%s`" % get_name_from_full_name(ref),
                                                     "#" + get_name_from_full_name(ref).lower())

            return "`Array<`%s`>`" % link("`%s`" % get_name_from_full_name(ref),
                                          "#" + get_name_from_full_name(ref).lower())

        return "`Array< %s >`" % get_name_from_full_name(field['items']['type'])

    elif obj_type_raw == "object":
        if '$ref' in field['additionalProperties']:
            ref = field['additionalProperties']['$ref'][14:]
            add_to_queue(ref, json_field_name)
            return link("`%s`" % get_name_from_full_name(ref), "#" + get_name_from_full_name(ref).lower())

        else:
            return "`Map< %s , %s >`" % get_key_value_field_types(field)

    elif 'format' in field:
        return "`%s`" % field['format']

    return "`%s`" % field['type']


class DocGeneratorContext:

    def __init__(self):
        self.defs = {}
        self.completed_fields = set()
        self.json_name = {}
        self.queue = ['io.argoproj.workflow.v1alpha1.Workflow', 'io.argoproj.workflow.v1alpha1.CronWorkflow',
                      'io.argoproj.workflow.v1alpha1.WorkflowTemplate']
        self.external = []
        self.index = {}

    def run(self):
        self.load_files()

        out = SECTION_HEADER % "Argo Fields"
        while self.queue:
            out += self.get_template(self.queue.pop(0))

        out += SECTION_HEADER % "External Fields"
        while self.external:
            out += self.get_template(self.external.pop(0))

        return out

    def load_files(self):
        with open('api/openapi-spec/swagger.json') as swagger:
            s = json.load(swagger)
            self.defs = s['definitions']

        for file_name_poisx in Path('examples').rglob('*.yaml'):
            file_name = str(file_name_poisx)
            with open(file_name) as wf:
                wf_text = wf.read()
                kinds = re.findall(r"kind: ([a-zA-Z]+)", wf_text)
                for kind in kinds:
                    if kind not in self.index:
                        self.index[kind] = set()
                    self.index[kind].add(file_name)

                finds = re.findall(r"([a-zA-Z]+?):", wf_text)
                for find in finds:
                    if find not in self.index:
                        self.index[find] = set()
                    self.index[find].add(file_name)

    def add_to_queue(self, ref, json_field_name):
        if ref == "io.argoproj.workflow.v1alpha1.ParallelSteps":
            ref = "io.argoproj.workflow.v1alpha1.WorkflowStep"
        if ref not in self.completed_fields:
            self.completed_fields.add(ref)
            self.json_name[ref] = json_field_name
            if "io.argoproj.workflow" in ref:
                self.queue.append(ref)
            else:
                self.external.append(ref)

    def get_template(self, key):
        out = FIELD_HEADER % (get_name_from_full_name(key), get_desc(key, self.defs))

        if get_name_from_full_name(key) in self.index:
            out += get_examples(self.index[get_name_from_full_name(key)], "Examples")

        if key in self.json_name and self.json_name[key] in self.index:
            out += get_examples(self.index[self.json_name[key]], "Examples with this field")

        if 'properties' not in self.defs[key]:
            return out

        out += FIELD_TABLE_HEADER
        for json_field_name, field in self.defs[key]['properties'].items():
            if '$ref' in field:

                ref = field['$ref'][14:]
                self.add_to_queue(ref, json_field_name)

                desc = get_desc_from_field(field)
                out += get_row(json_field_name,
                               link("`%s`" % get_name_from_full_name(ref), "#" + get_name_from_full_name(ref).lower()),
                               clean_desc(desc))

            else:
                obj_type = get_object_type(field, json_field_name, lambda r, j: self.add_to_queue(r, j))
                desc = get_desc_from_field(field)

                out += get_row(json_field_name, obj_type, clean_desc(desc))

        return out

# This file should be run with `make docs` on the root directory

ctx = DocGeneratorContext()
fields = ctx.run()
with open('docs/fields.md', 'w') as file:
    file.write(fields)

File: hack/docgen.go
//go:build !fields
// +build !fields

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

const sectionHeader = `

# %s
`

const fieldHeader = `

## %s

%s`

const fieldTableHeader = `

### Fields
| Field Name | Field Type | Description   |
|:----------:|:----------:|---------------|`

const tableRow = `
|` + "`%s`" + `|%s|%s|`

const depTableRow = `
|~~` + "`%s`" + `~~|~~%s~~|%s|`

// `markdown` attribute for MD in HTML: https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown/#markdown-in-html
const dropdownOpener = `

<details markdown>
<summary>%s (click to open)</summary>`

const listElement = `

- %s`

const dropdownCloser = `
</details>`

func cleanTitle(title string) string {
	if index := strings.Index(title, "+g"); index != -1 {
		return title[:index]
	}
	return title
}

func cleanDesc(desc string) string {
	desc = strings.ReplaceAll(desc, "\n", " ")
	desc = strings.ReplaceAll(desc, "  ", " ") // reduce multiple spaces to a single space
	dep := ""
	if index := strings.Index(desc, "DEPRECATED"); index != -1 {
		dep = " " + desc[:index]
	}

	if index := strings.Index(desc, "+patch"); index != -1 {
		desc = desc[:index]
	}
	if index := strings.Index(desc, "+proto"); index != -1 {
		desc = desc[:index]
	}
	if index := strings.Index(desc, "+option"); index != -1 {
		desc = desc[:index]
	}

	if dep != "" && !strings.Contains(desc, "DEPRECATED") {
		desc += dep
	}
	return desc
}

func getRow(name, objType, desc string) string {
	if index := strings.Index(desc, "DEPRECATED"); index != -1 {
		return fmt.Sprintf(depTableRow, name, objType, "~~"+desc[:index-1]+"~~ "+desc[index:])
	}
	return fmt.Sprintf(tableRow, name, objType, desc)
}

func getNameFromFullName(fullName string) string {
	split := strings.Split(fullName, ".")
	return split[len(split)-1]
}

func link(text, linkTo string) string {
	return fmt.Sprintf("[%s](%s)", text, linkTo)
}

func getDescFromField(field map[string]interface{}) string {
	if val, ok := field["description"]; ok {
		return cleanDesc(val.(string))
	} else if val, ok := field["title"]; ok {
		return cleanDesc(cleanTitle(val.(string)))
	}
	return "_No description available_"
}

func getExamples(examples Set, summary string) string {
	out := fmt.Sprintf(dropdownOpener, summary)
	for _, example := range sortedSetKeys(examples) {
		split := strings.Split(example, "/")
		name := split[len(split)-1]
		out += fmt.Sprintf(listElement, link(fmt.Sprintf("`%s`", name), "https://github.com/argoproj/argo-workflows/blob/main/"+example))
	}
	out += dropdownCloser
	return out
}

func getKeyValueFieldTypes(field map[string]interface{}) (string, string) {
	keyType, valType := "string", "string"
	addProps := field["additionalProperties"].(map[string]interface{})
	if val, ok := addProps["type"]; ok {
		keyType = val.(string)
	}
	if val, ok := addProps["format"]; ok {
		valType = val.(string)
	}
	return keyType, valType
}

func getObjectType(field map[string]interface{}, addToQueue func(string)) string {
	objTypeRaw := field["type"].(string)
	if objTypeRaw == "array" {
		if ref, ok := field["items"].(map[string]interface{})["$ref"]; ok {
			refString := ref.(string)[14:]
			addToQueue(refString)

			name := getNameFromFullName(refString)
			if refString == "io.argoproj.workflow.v1alpha1.ParallelSteps" {
				return fmt.Sprintf("`Array<Array<`%s`>>`", link(fmt.Sprintf("`%s`", "WorkflowStep"), fmt.Sprintf("#"+strings.ToLower("WorkflowStep"))))
			}
			return fmt.Sprintf("`Array<`%s`>`", link(fmt.Sprintf("`%s`", name), fmt.Sprintf("#"+strings.ToLower(name))))
		}
		fullName := field["items"].(map[string]interface{})["type"].(string)
		return fmt.Sprintf("`Array< %s >`", getNameFromFullName(fullName))
	} else if objTypeRaw == "object" {
		if ref, ok := field["additionalProperties"].(map[string]interface{})["$ref"]; ok {
			refString := ref.(string)[14:]
			addToQueue(refString)
			name := getNameFromFullName(refString)
			return link(fmt.Sprintf("`%s`", name), "#"+strings.ToLower(name))
		}
		key, val := getKeyValueFieldTypes(field)
		return fmt.Sprintf("`Map< %s , %s >`", key, val)
	} else if format, ok := field["format"].(string); ok {
		return fmt.Sprintf("`%s`", format)
	}
	return fmt.Sprintf("`%s`", field["type"].(string))
}

func glob(dir string, ext string) ([]string, error) {
	var files []string
	err := filepath.Walk(dir, func(path string, f os.FileInfo, err error) error {
		if filepath.Ext(path) == ext {
			files = append(files, path)
		}
		return nil
	})
	return files, err
}

func sortedMapInterfaceKeys(in map[string]interface{}) []string {
	var stringList []string
	for key := range in {
		stringList = append(stringList, key)
	}
	sort.Strings(stringList)
	return stringList
}

func sortedSetKeys(in Set) []string {
	var stringList []string
	for key := range in {
		stringList = append(stringList, key)
	}
	sort.Strings(stringList)
	return stringList
}

type DocGeneratorContext struct {
	doneFields Set
	queue      []string
	external   []string
	index      map[string]Set
	jsonName   map[string]string
	defs       map[string]interface{}
}

type Set map[string]bool

func NewDocGeneratorContext() *DocGeneratorContext {
	return &DocGeneratorContext{
		doneFields: make(Set),
		queue: []string{
			"io.argoproj.workflow.v1alpha1.Workflow", "io.argoproj.workflow.v1alpha1.CronWorkflow",
			"io.argoproj.workflow.v1alpha1.WorkflowTemplate",
		},
		external: []string{},
		index:    make(map[string]Set),
		jsonName: make(map[string]string),
		defs:     make(map[string]interface{}),
	}
}

func (c *DocGeneratorContext) loadFiles() {
	bytes, err := os.ReadFile("api/openapi-spec/swagger.json")
	if err != nil {
		panic(err)
	}
	swagger := make(map[string]interface{})
	err = json.Unmarshal(bytes, &swagger)
	if err != nil {
		panic(err)
	}
	c.defs = swagger["definitions"].(map[string]interface{})

	files, err := glob("examples/", ".yaml")
	if err != nil {
		panic(err)
	}
FILES:
	for _, fileName := range files {
		bytes, err := os.ReadFile(filepath.Clean(fileName))
		if err != nil {
			panic(err)
		}

		r := regexp.MustCompile(`kind: ([a-zA-Z]+)`)
		matches := r.FindAllStringSubmatch(string(bytes), -1)
		for _, m := range matches {
			kind := m[1]
			switch kind {
			case "ClusterWorkflowTemplate", "CronWorkflow", "Workflow", "WorkflowTemplate":
			default:
				continue FILES
			}
			if set, ok := c.index[kind]; ok {
				set[fileName] = true
			} else {
				c.index[kind] = make(Set)
				c.index[kind][fileName] = true
			}
		}

		r = regexp.MustCompile(`([a-zA-Z]+?):`)
		finds := r.FindAllStringSubmatch(string(bytes), -1)
		for _, find := range finds {
			if set, ok := c.index[find[1]]; ok {
				set[fileName] = true
			} else {
				c.index[find[1]] = make(Set)
				c.index[find[1]][fileName] = true
			}
		}
	}
}

func (c *DocGeneratorContext) addToQueue(ref, jsonFieldName string) {
	if ref == "io.argoproj.workflow.v1alpha1.ParallelSteps" {
		ref = "io.argoproj.workflow.v1alpha1.WorkflowStep"
	}
	if _, ok := c.doneFields[ref]; !ok {
		c.doneFields[ref] = true
		c.jsonName[ref] = jsonFieldName
		if strings.Contains(ref, "io.argoproj.workflow") {
			c.queue = append(c.queue, ref)
		} else {
			c.external = append(c.external, ref)
		}
	}
}

func (c *DocGeneratorContext) getDesc(key string) string {
	obj, ok := c.defs[key].(map[string]interface{})
	if !ok {
		return "_No description available_"
	}
	if val, ok := obj["description"]; ok {
		return cleanDesc(val.(string))
	} else if val, ok := obj["title"]; ok {
		return cleanDesc(cleanTitle(val.(string)))
	}
	return "_No description available_"
}

func (c *DocGeneratorContext) getTemplate(key string) string {
	name := getNameFromFullName(key)
	out := fmt.Sprintf(fieldHeader, name, c.getDesc(key))

	if set, ok := c.index[name]; ok {
		out += getExamples(set, "Examples")
	}
	if jsonName, ok := c.jsonName[key]; ok {
		if set, ok := c.index[jsonName]; ok {
			out += getExamples(set, "Examples with this field")
		}
	}

	var properties map[string]interface{}
	def, ok := c.defs[key]
	if !ok {
		return out
	}
	if props, ok := def.(map[string]interface{})["properties"]; ok {
		properties = props.(map[string]interface{})
	} else {
		return out
	}

	out += fieldTableHeader
	for _, jsonFieldName := range sortedMapInterfaceKeys(properties) {
		field := properties[jsonFieldName].(map[string]interface{})
		if ref, ok := field["$ref"]; ok {
			refString := ref.(string)[14:]
			c.addToQueue(refString, jsonFieldName)

			desc := getDescFromField(field)
			refName := getNameFromFullName(refString)
			out += getRow(jsonFieldName, link(fmt.Sprintf("`%s`", refName), "#"+strings.ToLower(refName)), cleanDesc(desc))
		} else {
			objType := getObjectType(field, func(ref string) { c.addToQueue(ref, jsonFieldName) })
			desc := getDescFromField(field)
			out += getRow(jsonFieldName, objType, cleanDesc(desc))
		}
	}
	return out
}

func (c *DocGeneratorContext) generate() string {
	c.loadFiles()

	out := "# Field Reference"
	for len(c.queue) > 0 {
		var temp string
		temp, c.queue = c.queue[0], c.queue[1:]
		out += c.getTemplate(temp)
	}

	out += fmt.Sprintf(sectionHeader, "External Fields")
	for len(c.external) > 0 {
		var temp string
		temp, c.external = c.external[0], c.external[1:]
		out += c.getTemplate(temp)
	}

	out += "\n"
	return out
}

func generateDocs() {
	println("generating docs/fields.md")
	c := NewDocGeneratorContext()
	err := os.WriteFile("docs/fields.md", []byte(c.generate()), 0o600)
	if err != nil {
		panic(err)
	}
}
